### PR TITLE
vhost_user: Add support for VHOST_USER_PROTOCOL_F_CONFIGURE_MEM_SLOTS

### PR DESCRIPTION
**Pull request by sboeuf**
This PR introduces the support for recently added message types `VHOST_USER_GET_MAX_MEM_SLOTS`, `VHOST_USER_ADD_MEM_REG` and `VHOST_USER_REM_MEM_REG`, coming from the latest vhost-user specification, and gated by the protocol feature `VHOST_USER_PROTOCOL_F_CONFIGURE_MEM_SLOTS`.

### CODE REVIEW

**Comment by sboeuf** [comment]
It cannot be merged until https://github.com/rust-vmm/vhost/pull/27 is merged, as it needs to be rebased once this happens.

**Comment by sboeuf** [comment]
@slp have you tried to move this crate to the latest vhost, and then update the `virtiofsd-rs` to rely on it?
I'm asking because I played with it, and unfortunately it's not working well when using DAX. I get a vCPU error from my guest, which I don't really know how to interpret.

**Comment by slp** [comment]
@sboeuf I tested it without DAX. Will try again with DAX tomorrow. Thanks for the heads up!

**Comment by slp** [comment]
@sboeuf Seems to be working here. Which error are you getting and in which context?

**Comment by sboeuf** [comment]
@slp so I was trying this with Cloud Hypervisor, and reading a small file was working fine (with DAX). But as soon as I ran `fio` with the following command:
```
sudo fio --filename=/home/cloud/mount_dir/test --size=2G --direct=1 --rw=randrw --bs=4k --ioengine=sync --iodepth=1 --runtime=60 --numjobs=1 --time_based --group_reporting --name=iops-test-job --eta-newline=1
```
it started failing with:
```
Starting 1 process                                                                                                                  
cloud-hypervisor: 23.837141048s: <vcpu0> ERROR:vmm/src/cpu.rs:945 -- VCPU generated error: VcpuRun(Failed to run vcpu: VCPU error Error(14)
                                                                                                                                    
Caused by:                                                                                                                          
    VCPU error Error(14))                                                                                                           
Killed                                                                                                                              
```

**Comment by slp** [comment]
@sboeuf You're right, it also happens with QEMU. The problem is that the previous implementation of `SlaveFsCacheReq` in `vhost` always waited for an ack from the VMM, but on the newer it must be explicitly enabled by calling `SlaveFsCacheReq::set_reply_ack_flag` if `VHOST_USER_PROTOCOL_F_REPLY_ACK` has been negotiated. The result is that the Guest may attempt to access the DAX window before the mapping has been set up by the VMM.

We need to update `VhostUserSlaveReqHandler::set_slave_req_fd` to do this, but we also need your `lib: Fix codebase with VhostUserSlaveReqHandlerMut` and another commit adding the new hot-[un]plug memory methods that have been added to the traits (`get_max_mem_slots`, `add_mem_region`, `remove_mem_region`).

I think  you were planning to add such methods with this PR. If so, please consider also adding one with the `set_reply_ack_flag` thingy, so it serves to bring up the whole crate up to date with `vhost`.

Thanks!

**Comment by sboeuf** [comment]
Oh cool! and thanks for checking this @slp 
I'm gonna add the missing bits to my PR and we should be good to go :)

**Comment by sboeuf** [comment]
@slp the PR is ready to be reviewed :)

**Comment by sboeuf** [comment]
I've tested this PR with Cloud Hypervisor and my `virtiofsd-rs` branch [here](https://gitlab.com/sboeuf/virtiofsd-rs/-/tree/update_vhost_backend), and it works like a charm :)

File: Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ license = "Apache-2.0"
 epoll = ">=4.0.1"
 libc = ">=0.2.39"
 log = ">=0.4.6"
-vhost_rs = { git = "https://github.com/rust-vmm/vhost", package = "vhost", features = ["vhost-user-slave"] }
+vhost = { git = "https://github.com/rust-vmm/vhost", package = "vhost", features = ["vhost-user-slave"] }
```
Comment (jiangliu) [inline Cargo.toml:12]: the package attribute could be skipped?

**Review by jiangliu — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (sboeuf) [reply Cargo.toml]: You're right, I removed it.

**Review by slp — APPROVED** [review]
After applying https://github.com/rust-vmm/vhost/pull/24 to `vhost`, and modifying `virtiofsd-rs` to return `VhostUserProtocolFeatures::REPLY_ACK` on `protocol_features()`, works fine with QEMU, including DAX.

**Comment by sboeuf** [comment]
@slp I've sent a PR for `virtiofsd-rs` https://gitlab.com/sboeuf/virtiofsd-rs/-/merge_requests/1

**Comment by slp** [comment]
@sboeuf I think the MR is in the wrong repo, it should be in https://gitlab.com/virtio-fs/virtiofsd-rs

**Comment by sboeuf** [comment]
Oops! Let me change that :)

**Comment by sboeuf** [comment]
@slp I replaced it with https://gitlab.com/virtio-fs/virtiofsd-rs/-/merge_requests/14

**Pull request merged**